### PR TITLE
Refactor AndroidContextGL, AndroidSurface and AndroidSurfaceGL

### PR DESCRIPTION
**Pull request by blasten**
This will make the relation between `AndroidContext` and `AndroidSurface` closer to `IOSContext` and `IOSSurface`.

With this change, every GL surface will own an onscreen and offscreen `EGLSurface` as opposed to the Android context owning the surfaces.
 
Fixes https://github.com/flutter/flutter/issues/58269

### CODE REVIEW

File: shell/platform/android/android_context.cc [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    : rendering_api_(rendering_api) {}
+
+AndroidContext::~AndroidContext() = default;
+
+AndroidRenderingAPI AndroidContext::RenderingApi() {
+  return rendering_api_;
+}
+
+bool AndroidContext::IsValid() {
+  return true;
```
Comment (chinmaygarde) [inline shell/platform/android/android_context.cc:22]: What is the point of this method if it is always true? It seems you are not returning a non-valid context from the factory. If you don't plan on adding to this, I'd get rid of this.

File: shell/platform/android/android_context.cc [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    case AndroidRenderingAPI::kOpenGLES:
+      context = std::make_shared<AndroidContextGL>(
+          rendering_api, fml::MakeRefCounted<AndroidEnvironmentGL>());
+      break;
+    case AndroidRenderingAPI::kVulkan:
+      context = std::make_shared<AndroidContext>(rendering_api);
+      break;
+  }
+  FML_CHECK(context);
+  FML_CHECK(context->IsValid()) << "Could not create an Android context.";
```
Comment (chinmaygarde) [inline shell/platform/android/android_context.cc:41]: IsValid is not virtual and always returns true. So this check is redundant. I suggest making IsValid a const pure virtual so that AndroidContext itself is abstract. Then in the IsValid override for each, you can return if  the context is actually valid.

File: shell/platform/android/android_context.cc [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+  std::shared_ptr<AndroidContext> context;
+  switch (rendering_api) {
+    case AndroidRenderingAPI::kSoftware:
+      context = std::make_shared<AndroidContext>(rendering_api);
+      break;
+    case AndroidRenderingAPI::kOpenGLES:
+      context = std::make_shared<AndroidContextGL>(
+          rendering_api, fml::MakeRefCounted<AndroidEnvironmentGL>());
+      break;
+    case AndroidRenderingAPI::kVulkan:
```
Comment (chinmaygarde) [inline shell/platform/android/android_context.cc:36]: This breaking Vulkan builds right? (its been quite a long time since I have checked those builds BTW, if maintaining those is getting in the way, we should get rid of them).

File: shell/platform/android/android_context.h [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+  static std::shared_ptr<AndroidContext> Create(
+      AndroidRenderingAPI rendering_api);
+
+  AndroidRenderingAPI RenderingApi();
+
+  bool IsValid();
+
+ private:
+  AndroidRenderingAPI rendering_api_;
```
Comment (chinmaygarde) [inline shell/platform/android/android_context.h:38]: Make this const so its impossible to forget to initialize this ivar.

File: shell/platform/android/android_context_gl.cc [hunk 5]
```diff
@@ -105,82 +105,41 @@ static bool TeardownContext(EGLDisplay display, EGLContext context) {
...
-
-  surface_ = eglCreatePbufferSurface(display, config_, attribs);
-  return surface_ != EGL_NO_SURFACE;
-}
-
-AndroidContextGL::AndroidContextGL(fml::RefPtr<AndroidEnvironmentGL> env,
-                                   const AndroidContextGL* share_context)
-    : environment_(env),
-      window_(nullptr),
+AndroidContextGL::AndroidContextGL(
```
Comment (chinmaygarde) [inline shell/platform/android/android_context_gl.cc:108]: Much cleaner. Thanks!

File: shell/platform/android/android_context_gl.cc [hunk 6]
```diff
@@ -196,13 +155,39 @@ AndroidContextGL::~AndroidContextGL() {
...
-        << "Could not tear down the EGL surface. Possible resource leak.";
+  if (!TeardownContext(environment_->Display(), resource_context_)) {
+    FML_LOG(ERROR) << "Could not tear down the EGL resource context. Possible "
+                      "resource leak.";
     LogLastEGLError();
   }
 }
 
+// For onscreen rendering.
+EGLSurface AndroidContextGL::CreateOnscreenSurface(
```
Comment (chinmaygarde) [inline shell/platform/android/android_context_gl.cc:166]: Prefer RAII objects. It is not clear from this method or its documentation about how and when the EGLSurface must be collected. See fml/unique_object.h for one possible utility to implement a unique EGLSurface that cleans up after itself.

Comment (chinmaygarde) [inline shell/platform/android/android_context_gl.cc:166]: Prefer RAII objects. It is not clear from this method or its documentation about how and when the EGLSurface must be collected. See fml/unique_object.h for one possible utility to implement a unique EGLSurface that cleans up after itself.

File: shell/platform/android/android_context_gl.cc [hunk 7]
```diff
@@ -196,13 +155,39 @@ AndroidContextGL::~AndroidContextGL() {
...
+    fml::RefPtr<AndroidNativeWindow> window) {
+  // The configurations are only required when dealing with extensions or VG.
```
Comment (chinmaygarde) [inline shell/platform/android/android_context_gl.cc:168]: What configurations? This comment  is now out of date since you moved the config_ initialization maybe?

File: shell/platform/android/android_context_gl.cc [hunk 8]
```diff
@@ -234,41 +229,29 @@ bool AndroidContextGL::ClearCurrent() {
...
 
-SkISize AndroidContextGL::GetSize() {
+SkISize AndroidContextGL::GetSize(EGLSurface surface) {
   EGLint width = 0;
   EGLint height = 0;
 
-  if (!eglQuerySurface(environment_->Display(), surface_, EGL_WIDTH, &width) ||
-      !eglQuerySurface(environment_->Display(), surface_, EGL_HEIGHT,
-                       &height)) {
+  if (!eglQuerySurface(environment_->Display(), surface, EGL_WIDTH, &width) ||
```
Comment (chinmaygarde) [inline shell/platform/android/android_context_gl.cc:241]: If you are going to make EGLSurface an RAII object, it may make sense to put these behind accessors. Just a thought.

File: shell/platform/android/android_context_gl.h [hunk 9]
```diff
@@ -9,47 +9,49 @@
...
 #include "flutter/shell/common/platform_view.h"
+#include "flutter/shell/platform/android/android_context.h"
 #include "flutter/shell/platform/android/android_environment_gl.h"
 #include "flutter/shell/platform/android/android_native_window.h"
 #include "third_party/skia/include/core/SkSize.h"
 
 namespace flutter {
 
-class AndroidContextGL : public fml::RefCountedThreadSafe<AndroidContextGL> {
+class AndroidContextGL : public AndroidContext {
```
[the remark this answers is not in the document]
Comment (chinmaygarde) [reply shell/platform/android/android_context_gl.h:19]: Sound good.

File: shell/platform/android/android_context_gl.h [hunk 10]
```diff
@@ -9,47 +9,49 @@
...
  public:
-  bool CreateWindowSurface(fml::RefPtr<AndroidNativeWindow> window);
+  AndroidContextGL(AndroidRenderingAPI rendering_api,
+                   fml::RefPtr<AndroidEnvironmentGL> environment);
 
-  bool CreatePBufferSurface();
+  ~AndroidContextGL();
+
+  EGLSurface CreateOnscreenSurface(fml::RefPtr<AndroidNativeWindow> window);
```
Comment (chinmaygarde) [inline shell/platform/android/android_context_gl.h:26]: These should be const IMO.

File: shell/platform/android/android_context_gl.h [hunk 11]
```diff
@@ -9,47 +9,49 @@
...
+
+  EGLSurface CreateOffscreenSurface();
+
+  void SetEnvironment(fml::RefPtr<AndroidEnvironmentGL>);
```
Comment (chinmaygarde) [inline shell/platform/android/android_context_gl.h:30]: Why is this necessary? I don't see it being used anywhere.

File: shell/platform/android/android_context_gl.h [hunk 12]
```diff
@@ -9,47 +9,49 @@
...
-  bool CreateWindowSurface(fml::RefPtr<AndroidNativeWindow> window);
+  AndroidContextGL(AndroidRenderingAPI rendering_api,
+                   fml::RefPtr<AndroidEnvironmentGL> environment);
 
-  bool CreatePBufferSurface();
+  ~AndroidContextGL();
+
+  EGLSurface CreateOnscreenSurface(fml::RefPtr<AndroidNativeWindow> window);
+
+  EGLSurface CreateOffscreenSurface();
```
Comment (chinmaygarde) [inline shell/platform/android/android_context_gl.h:28]: Here and elsewhere, doxygen style docs please. Here I would mention this creates a 1x1 pbuffer surface used for making the offscreen context current for texture uploads.

File: shell/platform/android/android_context.cc (hunk 1 above)
Comment (blasten) [reply shell/platform/android/android_context.cc]: Done. This has been removed.

File: shell/platform/android/android_context.cc (hunk 2 above)
Comment (blasten) [reply shell/platform/android/android_context.cc]: Done.

File: shell/platform/android/android_context.cc (hunk 3 above)
Comment (blasten) [reply shell/platform/android/android_context.cc]: It shouldn't be a break the build AFAICT.

File: shell/platform/android/android_context.h (hunk 4 above)
Comment (blasten) [reply shell/platform/android/android_context.h]: Done.

File: shell/platform/android/android_context_gl.h (hunk 12 above)
Comment (blasten) [reply shell/platform/android/android_context_gl.h:53]: Done.

File: shell/platform/android/android_context_gl.h (hunk 11 above)
Comment (blasten) [reply shell/platform/android/android_context_gl.h]: Removed.

File: shell/platform/android/android_context_gl.h (hunk 10 above)
Comment (blasten) [reply shell/platform/android/android_context_gl.h]: Done.

File: shell/platform/android/android_context_gl.cc (hunk 8 above)
Comment (blasten) [reply shell/platform/android/android_context_gl.cc:235]: I tried to use ` fml/unique_object.h`, but I realized it was a bit tricky for one reason:
To destroy an `EGLSurface`, an instance of EGLDisplay is required. This is currently held by `AndroidEnvironmentGL`.  While we currently just use `EGL_DEFAULT_DISPLAY`, I shouldn't assume that in a `Free(EGLSurface* surface)`  method.

**Comment by blasten** [comment]
PTAL

File: shell/platform/android/android_context_gl.cc (hunk 8 above)
Comment (chinmaygarde) [reply shell/platform/android/android_context_gl.cc:235]: Discussed this via chat.
